File: ai/src/main/java/org/pingel/gestalt/core/Traversal.java
package org.pingel.gestalt.core;

public class Traversal
{
    private String offset;
    private char[] chars;
    
    public Traversal(String n) 
    {
		offset = n;
		chars = offset.toCharArray();
    }

    public String toString()
    {
        return offset;
    }

    public Traversal copy()
    {
		return new Traversal(offset);
    }

    public void append(String tail)
    {
		offset += tail;
		chars = offset.toCharArray();
    }

    public int length()
    {
		return chars.length;
    }

    public char elementAt(int i)
    {
		return chars[i];
    }

}

File: ai/src/main/scala/org/pingel/gestalt/core/Traversal.scala
package org.pingel.gestalt.core

case class Traversal(var offset: String)
{
  var chars = offset.toList

  def toString = offset

  def copy() = new Traversal(offset)

  def append(tail: String): Unit = {
    offset += tail
    chars = offset.toList
  }

  def length() = chars.length

  def elementAt(i: Int) = chars(i)

}
